File: flink-app/src/main/java/com/example/ctr/config/AppConfig.java
package com.example.ctr.config;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import lombok.Getter;
import lombok.Setter;

import java.io.InputStream;

@Getter
@Setter
public class AppConfig {

    private KafkaProperties kafka;
    private RedisProperties redis;
    private ClickHouseProperties clickhouse;
    private DuckDBProperties duckdb;
    private CtrConfig ctr;
    private FlinkConfig flink;

    @Getter
    @Setter
    public static class CtrConfig {
        private CtrJobProperties job;
    }

    @Getter
    @Setter
    public static class FlinkConfig {
        private int parallelism;
        private CheckpointConfig checkpoint;
        private StateBackendConfig stateBackend;
        private RestartStrategyConfig restartStrategy;

        @Getter
        @Setter
        public static class CheckpointConfig {
            private boolean enabled;
            private long interval;
            private long timeout;
            private long minPause;
            private int maxConcurrent;
            private String mode;
        }

        @Getter
        @Setter
        public static class StateBackendConfig {
            private String type;
            private String checkpointDir;
            private String savepointDir;
        }

        @Getter
        @Setter
        public static class RestartStrategyConfig {
            private String type;
            private int attempts;
            private long delay;
        }
    }

    public static AppConfig load() {
        try {
            ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
            mapper.setPropertyNamingStrategy(PropertyNamingStrategies.KEBAB_CASE);

            // Try loading application-local.yml first, then application.yml
            InputStream is = AppConfig.class.getClassLoader().getResourceAsStream("application-local.yml");
            if (is == null) {
                is = AppConfig.class.getClassLoader().getResourceAsStream("application.yml");
            }
            if (is == null) {
                throw new RuntimeException("Configuration file not found (application-local.yml or application.yml)");
            }

            return mapper.readValue(is, AppConfig.class);
        } catch (Exception e) {
            throw new RuntimeException("Failed to load configuration", e);
        }
    }
}

File: flink-app/src/main/kotlin/com/example/ctr/config/AppConfig.kt
package com.example.ctr.config

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import java.io.InputStream

@JsonIgnoreProperties(ignoreUnknown = true)
class AppConfig {

    var kafka: KafkaProperties = KafkaProperties()
    var redis: RedisProperties = RedisProperties()
    var clickhouse: ClickHouseProperties = ClickHouseProperties()
    var duckdb: DuckDBProperties = DuckDBProperties()
    var ctr: CtrConfig = CtrConfig()
    var flink: FlinkConfig = FlinkConfig()

    class CtrConfig {
        var job: CtrJobProperties = CtrJobProperties()
    }

    class FlinkConfig {
        var parallelism: Int = 1
        var checkpoint: CheckpointConfig = CheckpointConfig()
        var stateBackend: StateBackendConfig = StateBackendConfig()
        var restartStrategy: RestartStrategyConfig = RestartStrategyConfig()

        class CheckpointConfig {
            var enabled: Boolean = true
            var interval: Long = 60_000L
            var timeout: Long = 600_000L
            var minPause: Long = 5_000L
            var maxConcurrent: Int = 1
            var mode: String = "EXACTLY_ONCE"
        }

        class StateBackendConfig {
            var type: String = "filesystem"
            var checkpointDir: String = ""
            var savepointDir: String = ""
        }

        class RestartStrategyConfig {
            var type: String = "fixed-delay"
            var attempts: Int = 3
            var delay: Long = 10_000L
        }
    }

    companion object {
        fun load(): AppConfig {
            try {
                val mapper = ObjectMapper(YAMLFactory())
                mapper.propertyNamingStrategy = PropertyNamingStrategies.KEBAB_CASE
                mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

                var inputStream: InputStream? = AppConfig::class.java.classLoader
                    .getResourceAsStream("application-local.yml")
                if (inputStream == null) {
                    inputStream = AppConfig::class.java.classLoader.getResourceAsStream("application.yml")
                }
                if (inputStream == null) {
                    throw RuntimeException("Configuration file not found (application-local.yml or application.yml)")
                }

                return mapper.readValue(inputStream, AppConfig::class.java)
            } catch (ex: Exception) {
                throw RuntimeException("Failed to load configuration", ex)
            }
        }
    }
}
